well additional information testing/implementation updates

jacob-a-brown opened the pull request:
This PR implements model/schema updates to pass all well additional information tests except for the aquifer/geology tests. Those will get implemented after the models have been finalized.

@chasetmartin:

- the feature step the response should include the source of the completion information seems to apply to the completion date of the well, is that correct?
- the feature step the response should include the source of the construction information seems to apply to the construction method of the well, is that correct?
- should other fields, like the driller name, also have a source?

jacob-a-brown committed:
refactor: rename Permission -> PermissionHistory | move mixin to same file

jacob-a-brown committed:
feat: add permission types to lexicon

jacob-a-brown committed:
refactor: import PermissionHistoryMixin from correct dir

jacob-a-brown committed:
fix: fix imports for newly renamed PermissionHistory

jacob-a-brown committed:
feat: add permission_type to lexicon

jacob-a-brown committed:
feat: update util to correspond with bdms 221

jacob-a-brown committed:
fix: sync bdms 221/227 for util functions

jacob-a-brown committed:
feat: implement permissions

jacob-a-brown committed:
feat: add is_suitable_for_datalogger to thing

jacob-a-brown committed:
Merge branch 'bdms-227' into bdms-227-jab-updates-to-pass-tests

jacob-a-brown committed:
fix: fix artifacts from merge conflicts

jacob-a-brown committed:
feat: implement data source tests for well additional information

jacob-a-brown committed:
fix: remove outdated notes

jacob-a-brown commented:
Besides the `completion date` the only construction information we have is the `construction method` (except for the construction notes). It seems to me that `DataSource` was kind of compound/nebulous and applied to both of these. Since the completion source step applies to the date I figured that the construction source would apply to the method. Do you think that would be an okay way to proceed? If so, should the feature files be changed to reflect this?

jacob-a-brown committed:
feat: implement origin_type

jacob-a-brown committed:
Merge branch 'bdms-227' into bdms-227-jab-updates-to-pass-tests

jacob-a-brown committed:
fix: fix __table_args__ for aquifer and geology

jacob-a-brown committed:
fix: add geologic_formation relation to WellScreen

This was missing on the WellScreen side of the relationship

jacob-a-brown committed:
fix: add missing comma

jacob-a-brown committed:
fix: use BaseResponseModel for GeoJSONProperties in location schema

jacob-a-brown committed:
fix: include baseresponseinfo in location geojson properties

jacob-a-brown committed:
feat: implement aquifer_systems in thing response

jacob-a-brown committed:
feat: fix aquifer systems test

jacob-a-brown committed:
fix: return aquifer system and types in dict

this allows us to couple types to a system incase a well is in
multiple aquifers

jacob-a-brown committed:
feat: implement geologic_formations from feature files

jacob-a-brown commented:
@ksmuczynski I renamed `aquifers` to `aquifer_systems` and `formations` to `geologic_formations` in the `thing` model `db/thing.py` to correspond with the table names and make it more explicit. I did this (for the former in particular) because the `ThingResponse` has a field called `aquifers` that concatenates `aquifer_systems` and `aquifer_types`

jirhiker commented on db/permission_history.py:
```diff
@@ -0,0 +1,96 @@
+            primaryjoin=(
+                and_(
+                    cls.id == foreign(PermissionHistory.target_id),
+                    PermissionHistory.target_table == pascal_to_snake(cls.__name__),
```
Can you use cls.__tablename__ here instead of having to convert __name__ yourself?

jacob-a-brown replied:
I'll try that out. If it works I'll adopt that standard for all of the polymorphic tables.

jacob-a-brown replied:
it works has been adopted for all polymorphic tables

jirhiker commented on tests/features/environment.py:
```diff
@@ -391,6 +514,46 @@ def before_all(context):
+        add_well_casing_material(context, session, well_1)
+
+        contact = add_contact(context, session)
+        add_permission_history(
```
Use a should be a loop to add the permission histories

jacob-a-brown committed:
refactor: clean up permission history testing data

use a for loop for cleaner and more maintainable code

jacob-a-brown committed:
fix: use origin_type to retrieve well depth source

jacob-a-brown committed:
refactor: use __tabename__ for polymorphic mixins

this ensures fidelity in table naming across all database models

jacob-a-brown committed:
fix: use __tablename__ for NotesMixin add_note

Now that __tablename__ is used throughout it should be
used for the target_table in add_note

jacob-a-brown committed:
refactor: use sets in feature tests for comparison

jacob-a-brown committed:
refactor: return permission history records for a well

instead of a specific permission type, return a list of permission records for all permission types associated with the well.

ksmuczynski reviewed (approved):
Looks good!
